core/src/main/java/org/elasticsearch/search/aggregations/pipeline/bucketsort/BucketSortPipelineAggregator.java

```diff
@@ -0,0 +1,194 @@
+        // Thus, we just have to pop as many elements as we expect in results and store them in reverse order.
+        LinkedList<InternalMultiBucketAggregation.InternalBucket> newBuckets = new LinkedList<>();
+        for (int i = resultSize - 1; i >= 0; --i) {
+            ComparableBucket comparableBucket = ordered.pop();
+            if (comparableBucket.skip == false) {
+                newBuckets.addFirst(comparableBucket.getInternalBucket());
+            }
+        }
+        return originalAgg.create(newBuckets);
+    }
+
+    private static List<InternalMultiBucketAggregation.InternalBucket> truncate(
+            List<? extends InternalMultiBucketAggregation.InternalBucket> buckets, int offset, int size) {
+
+        List<InternalMultiBucketAggregation.InternalBucket> truncated = new ArrayList<>(size);
+        for (int i = offset; i < offset + size; ++i) {
+            truncated.add(buckets.get(i));
+        }
+        return truncated;
+    }
+
+    private class ComparableBucket implements Comparable<ComparableBucket> {
+
+        private final MultiBucketsAggregation parentAgg;
+        private final InternalMultiBucketAggregation.InternalBucket internalBucket;
+
+        /**
+         * Whether the bucket should be skipped due to the gap policy
+         */
+        private boolean skip = false;
+
+        private ComparableBucket(MultiBucketsAggregation parentAgg, InternalMultiBucketAggregation.InternalBucket internalBucket) {
+            this.parentAgg = parentAgg;
+            this.internalBucket = internalBucket;
+        }
+
+        private InternalMultiBucketAggregation.InternalBucket getInternalBucket() {
+            return internalBucket;
+        }
+
+        @Override
+        public int compareTo(ComparableBucket that) {
+            for (FieldSortBuilder sort : sorts) {
+                String sortField = sort.getFieldName();
+                int compareResult = "_key".equals(sortField) ? compareKeys(this, that) : comparePathValues(sortField, this, that);
+                if (compareResult != 0) {
+                    return sort.order() == SortOrder.DESC ? compareResult : -compareResult;
+                }
+            }
+            return 0;
+        }
+
+        private int compareKeys(ComparableBucket b1, ComparableBucket b2) {
+            Comparable<Object> b1Key = (Comparable<Object>) b1.internalBucket.getKey();
+            Comparable<Object> b2Key = (Comparable<Object>) b2.internalBucket.getKey();
+            return b1Key.compareTo(b2Key);
+        }
+
+        private int comparePathValues(String sortField, ComparableBucket b1, ComparableBucket b2) {
+            Double b1Value = BucketHelpers.resolveBucketValue(parentAgg, b1.internalBucket, sortField, gapPolicy);
+            Double b2Value = BucketHelpers.resolveBucketValue(parentAgg, b2.internalBucket, sortField, gapPolicy);
```

Comment:
I wonder if we should just preemptively cache the bucket's value (and set skip) in the ctor for the `ComparableBucket`?  `resolveBucketValue()` isn't terribly expensive or anything, but since this happens during comparison, every push/pop onto the priority queue's heap will invoke a bunch of comparisons and re-resolving these values.

The downside is that the cached values may not be needed if we're only sorting on keys.  Dunno, thoughts?  

Might be affected by how/if the above skipping thing is actually an issue or not.